File: src/Polls/LJPollScaleQuestion.m
//
//  LJPollScaleQuestion.m
//  PollTest
//
//  Created by Fraser Speirs on Fri Mar 14 2003.
//  Copyright (c) 2003 Connected Flow. All rights reserved.
//

#import "LJPollScaleQuestion.h"

#define kPollScaleStart @"LJPollScaleStart"
#define kPollScaleEnd @"LJPollScaleEnd"
#define kPollScaleStep @"LJPollScaleStep"

@implementation LJPollScaleQuestion
@synthesize start;
@synthesize end;
@synthesize step;

+ (LJPollScaleQuestion *)scaleQuestionWithStart: (NSInteger)theStart end: (NSInteger)theEnd step:(NSInteger)theStep
{
    LJPollScaleQuestion *scale = [[LJPollScaleQuestion alloc] initWithStart:theStart end:theEnd step:theStep];

    return scale;
}

- (instancetype)init
{
    return [self initWithStart:0 end:100 step:10];
}

- (instancetype)initWithStart: (NSInteger)theStart end: (NSInteger)theEnd step:(NSInteger)theStep
{
    if (self = [super init]) {
        start = theStart;
        end = theEnd;
        step = theStep;
    }
    return self;
}

- (NSString *)htmlRepresentation
{
    return [NSString stringWithFormat: @"<lj-pq type=\"scale\" from=\"%ld\" to=\"%ld\" by=\"%ld\">%@</lj-pq>", (long)start, (long)end, (long)step, self.question];
}

#pragma mark Memento Pattern
- (NSDictionary *) memento
{
    NSDictionary *dict = @{kLJPollQuestionKey   : self.question,
                           kPollScaleStart      : @(start),
                           kPollScaleEnd        : @(end),
                           kPollScaleStep       : @(step)};

    return dict;
}

- (void) restoreFromMemento: (NSDictionary *)memento
{
    self.question = memento[kLJPollQuestionKey];
    self.start = [memento[kPollScaleStart] integerValue];
    self.end = [memento[kPollScaleEnd] integerValue];
    self.step = [memento[kPollScaleStep] integerValue];
}

#pragma mark NSCoding
- (void)encodeWithCoder:(NSCoder *)encoder
{
    [super encodeWithCoder: encoder];
    [encoder encodeInteger: start forKey: kPollScaleStart];
    [encoder encodeInteger: end forKey: kPollScaleEnd];
    [encoder encodeInteger: step forKey: kPollScaleStep];
}

- (instancetype)initWithCoder:(NSCoder *)decoder
{
    if (self = [super initWithCoder: decoder]) {
        start = [decoder decodeIntegerForKey: kPollScaleStart];
        end = [decoder decodeIntegerForKey: kPollScaleEnd];
        step = [decoder decodeIntegerForKey: kPollScaleStep];
    }
    return self;
}
@end

File: src/Polls/LJPollScaleQuestion.swift
//
//  LJPollScaleQuestion.swift
//  Xjournal
//
//  Created by C.W. Betts on 1/24/15.
//
//

import Foundation
import SwiftAdditions

private let kPollScaleStart = "LJPollScaleStart"
private let kPollScaleEnd = "LJPollScaleEnd"
private let kPollScaleStep = "LJPollScaleStep"

/// Question subclass representing a scale with start, end and step values
@objc(LJPollScaleQuestion) final class PollScaleQuestion: PollQuestion {
	dynamic var start: Int
	dynamic var end: Int
	dynamic var step: Int
	
	/// Returns a scale question with the given start, end and step values
	/// and a default question string
	init(start aStart: Int, end anEnd: Int, step aStep: Int) {
		start = aStart
		end = anEnd
		step = aStep
		
		super.init()
	}
	
	/// Returns a scale question within the given range, step values
	/// and a default question string
	convenience init(range: Range<Int>, step aStep: Int) {
		self.init(start: range.lowerBound, end: range.upperBound, step: aStep)
	}
	
	/// Returns a scale question within the given range, step values
	/// and a default question string
	convenience init(range: NSRange, step aStep: Int) {
		self.init(start: range.location, end: range.max, step: aStep)
	}
	
	/// Returns a scale question with a start of 1, an end of 100, a step of 10,
	/// and a default question string
	convenience override init() {
		self.init(start: 1, end: 100, step: 10)
	}
	
	/// Get the HTML representation
	override var htmlRepresentation: String {
		return "<lj-pq type=\"scale\" from=\"\(start)\" to=\"\(end)\" by=\"\(step)\">\(question)</lj-pq>"
	}
	
	// MARK: - memento
	override var memento: [String: Any] {
		return [kLJPollQuestionKey: question,
			kPollScaleStart: start,
			kPollScaleEnd: end,
			kPollScaleStep: step]
	}
	
	override func restore(fromMemento amemento: [String: Any]) {
		question = amemento[question] as! String
		start = amemento[kPollScaleStart] as! Int
		end = amemento[kPollScaleEnd] as! Int
		step = amemento[kPollScaleStep] as! Int
	}
	
	// MARK: - NSCoding
	override func encode(with aCoder: NSCoder) {
		super.encode(with: aCoder)
		
		aCoder.encode(start, forKey: kPollScaleStart)
		aCoder.encode(end, forKey: kPollScaleEnd)
		aCoder.encode(step, forKey: kPollScaleStep)
	}
	
	required init?(coder aDecoder: NSCoder) {
		start = aDecoder.decodeInteger(forKey: kPollScaleStart)
		end = aDecoder.decodeInteger(forKey: kPollScaleEnd)
		step = aDecoder.decodeInteger(forKey: kPollScaleStep)
		
		super.init(coder: aDecoder)
	}
	
	// MARK: -
	
	class var keyPathsForValuesAffectingHtmlRepresentation: Set<String> {
		return Set(["start", "end", "step", "question"])
	}
}
